### PR TITLE
KAFKA-3060: Refactor MeteredStore and RockDBStore Impl

### PR DESCRIPTION
**Pull request by guozhangwang**
Changes include:

1) Move logging logic from MeteredXXXStore to internal stores, and leave WindowedStore API clean by removed all internalPut/Get functions.

2) Wrap common logging behavior of InMemory and LRUCache stores into one class.

3) Fix a bug for StoreChangeLogger where byte arrays are not comparable in HashSet by using a specified RawStoreChangeLogger.

4) Add a caching layer on top of RocksDBStore with object caching, it relies on the object's equals and hashCode function to be consistent with serdes.

### CODE REVIEW

**Comment by guozhangwang** [comment]
Unit test passed locally.

**Comment by guozhangwang** [comment]
@ymatsuda Could you take a look?

**Comment by guozhangwang** [comment]
Rebased from trunk

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 1]
```diff
@@ -157,6 +157,7 @@
                                         ConsumerConfig.VALUE_DESERIALIZER_CLASS_DOC)
                                 .define(TIMESTAMP_EXTRACTOR_CLASS_CONFIG,
                                         Type.CLASS,
+                                        WALLCLOCK_TIMESTAMP_EXTRACTOR,
```
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:160]: Sneaky. This has nothing to do with this refactoring.
btw, did we agree on having this as the default timestamp extractor?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreSupplier.java [hunk 2]
```diff
@@ -43,13 +44,13 @@
 public class InMemoryKeyValueStoreSupplier<K, V> implements StateStoreSupplier {
 
     private final String name;
-    private final Serdes serdes;
     private final Time time;
+    private Serdes<K, V> serdes;
```
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreSupplier.java]: why not final?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/OffsetCheckpoint.java [hunk 3]
```diff
@@ -100,7 +100,7 @@ private void writeEntry(BufferedWriter writer, TopicPartition part, long offset)
 
     public Map<TopicPartition, Long> read() throws IOException {
         synchronized (lock) {
-            BufferedReader reader = null;
+            BufferedReader reader;
```
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/OffsetCheckpoint.java]: reader needs to be set to null here for the null check in the finally clause.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/OffsetCheckpoint.java [hunk 4]
```diff
@@ -136,8 +136,7 @@ private void writeEntry(BufferedWriter writer, TopicPartition part, long offset)
                         throw new IllegalArgumentException("Unknown offset checkpoint version: " + version);
                 }
             } finally {
-                if (reader != null)
-                    reader.close();
+                reader.close();
```
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/OffsetCheckpoint.java]: Restore the null check. Don't assume reader is always set.

Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/OffsetCheckpoint.java:139]: If the code path reached here, the reader should always be set?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreSupplier.java (hunk 2 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryKeyValueStoreSupplier.java]: ack.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 1 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:160]: @ymatsuda I thought it was actually merged in a previous PR for improving APIs, but obviously it gets left and slipped until now.

Let's discuss about this, I think your hesitation is that it should better to be a use's conscious decision which timestamp to use. Personally I would rather prefer to not enforce every user to deliberate on timestamps: generally speaking, I want to minimize the number of conscious decisions, i.e. required configs for users who use KStream for the first time, and ideally bootstrap brokers is the only thing you need to specify for your first KStream program. And as users start to use more they will dive into the details.

Thoughts?

**Comment by guozhangwang** [comment]
Replied / addressed comments.

**Comment by guozhangwang** [comment]
Unit test passed locally.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java [hunk 5]
```diff
@@ -17,25 +17,27 @@
...
-import org.apache.kafka.common.serialization.Deserializer;
-import org.apache.kafka.common.utils.Time;
 import org.apache.kafka.streams.state.KeyValueIterator;
 import org.apache.kafka.streams.state.KeyValueStore;
-import org.apache.kafka.streams.state.Serdes;
 
 import java.util.List;
 
+/**
+ * Metered KeyValueStore wrapper is used for both recording operation metrics and Kafka based logging purposes, and hence its
```
Comment (ewencp) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java]: This doesn't seem accurate -- this class no longer provides logging, it seems like maybe this has moved to `InMemoryKeyValueLoggedStore`.

**Comment by ewencp** [comment]
@guozhangwang LGTM, only a minor comment. But are we also ok with that change to the default config for `TIMESTAMP_EXTRACTOR_CLASS_CONFIG`? There didn't seem to be a conclusion to the discussion with @ymatsuda.

**Comment by guozhangwang** [comment]
@ymatsuda Please let me know if you still have concerns about `TIMESTAMP_EXTRACTOR_CLASS_CONFIG`?

**Comment by ewencp** [comment]
Ok, LGTM and there seems to be agreement on the `TIMESTAMP_EXTRACTOR_CLASS_CONFIG` default.

**Pull request closed without merging**